2020 JavaScript edits (#1737)
HTTPArchive/almanac.httparchive.org: src/content/en/2020/javascript.md

discuss: 2038
results: https://docs.google.com/spreadsheets/d/1cgXJrFH02SHPKDGD0AelaXAdB3UI7PIb5dlS0dxVtfY/
queries: 02_JavaScript

> featured_quote: JavaScript has come a long way from its humble origins as the last of the three web cornerstones—alongside CSS and HTML. Today, JavaScript has started to infiltrate a broad spectrum of the technical stack. JavaScript is no longer confined to the client-side—it's an increasingly popular choice for build tools and server-side scripting, and is creeping its way into the CDN layer as well thanks to edge computing solutions.

Reviewer: * Adding space around `—` in `... client-side — it's an ...`. Alternative, this can be split into two sentences.
* Should the `thanks to edge computing solutions` phrase be put in brackets separated by a comma or something?

```suggestion
featured_quote: JavaScript has come a long way from its humble origins as the last of the three web cornerstones—alongside CSS and HTML. Today, JavaScript has started to infiltrate a broad spectrum of the technical stack. JavaScript is no longer confined to the client-side — it's an increasingly popular choice for build tools and server-side scripting, and is creeping its way into the CDN layer as well thanks to edge computing solutions.
```

Author: We haven't been doing that in other chapters so would prefer to keep consistent.

Reviewer 2: The [style guide we are using says no spaces around em-dash](https://developers.google.com/style/dashes).

However I do agree it looks weird here so close to a hyphenated word.

Also I think this sentence is rather long with a first part, the em-dash, and then the third part. And the two "ands" don't help (though I agree technically correct).

I think we should rephrase. What about something like this?

```suggestion
featured_quote: JavaScript has come a long way from its humble origins as the last of the three web cornerstones—alongside CSS and HTML. Today, JavaScript has started to infiltrate a broad spectrum of the technical stack–it is no longer confined to the client-side and it's an increasingly popular choice for build tools and server-side scripting. JavaScript is also creeping its way into the CDN layer as well thanks to edge computing solutions.
```

Though not made about three sentences basically all starting with the same word (JavaScript), but that's probably just getting picky.

Reviewer: Look good to me, but I think we should even split it at `stack–it`. That is literally a hyphenated word now, which doesn't make sense.

Reviewer 2: Works for me, and also means we have a 4th sentence that doesn't begin with JavaScript 🙂

```suggestion
featured_quote: featured_quote: JavaScript has come a long way from its humble origins as the last of the three web cornerstones—alongside CSS and HTML. Today, JavaScript has started to infiltrate a broad spectrum of the technical stack. It is no longer confined to the client-side and it's an increasingly popular choice for build tools and server-side scripting. JavaScript is also creeping its way into the CDN layer as well thanks to edge computing solutions.
```

Author: Thoughts/suggestions @tkadlec?

featured_stat_1: 1,897ms
featured_stat_label_1: Median JS main thread time on mobile
featured_stat_2: 37.22%
featured_stat_label_2: Percentage of unused JS on mobile
featured_stat_3: 12.2%
featured_stat_label_3: Percentage of scripts loaded asynchronously
---

## Introduction

> JavaScript has come a long way from its humble origins as the last of the three web cornerstones, alongside CSS and HTML. Today, JavaScript has started to infiltrate a broad spectrum of the technical stack. JavaScript is no longer confined to the client-side—it's an increasingly popular choice for build tools and server-side scripting, and is creeping its way into the CDN layer as well thanks to edge computing solutions.

Reviewer: Same changes as above in the `featured_quote`.

) }}

## How do we load our JavaScript?

The way we load JavaScript has a significant impact on the overall experience.

By default, JavaScript is _parser-blocking_. In other words, when the browser discovers a `script` element, it must pause parsing of the HTML until the script has been downloaded, parsed, and executed. It's a significant bottleneck and a common contributor to pages that are slow to render.

> We can start to offset some of the cost of loading JavaScript by loading scripts either asynchronously, which only halts the HTML parser during the parse and execution phases and not during the download phase, or deferred, which doesn't halt the HTML parser at all. Both attributes are only available on external scripts—inline scripts cannot have them applied.

Reviewer: ```suggestion
We can start to offset some of the cost of loading JavaScript by loading scripts either asynchronously (with the `async` attribute`), which only halts the HTML parser during the parse and execution phases and not during the download phase, or deferred (with the `defer` attribute`), which doesn't halt the HTML parser at all. Both attributes are only available on external scripts — inline scripts cannot have them applied.
```

On mobile, external scripts comprise 59.0% of all script elements found.

<p class="note">

>   As an aside, when we talk about how much JavaScript is loaded on a page earlier, that total doesn't account for the size of these inline scripts—because they're part of the HTML document, they're counted against the markup size. This means we load even more script that the numbers show.

Reviewer: ```suggestion
  As an aside, when we talked about how much JavaScript is loaded on a page earlier, that total didn't account for the size of these inline scripts — because they're part of the HTML document, they're counted against the markup size. This means we load even more script that the numbers show.
```

  sql_file="breakdown_of_scripts_using_async_defer_module_nomodule.sql"
) }}

> {# TODO(authors): Why is it misleading? #}

Author: @tkadlec

Reviewer 3: Misleading was the wrong word choice. inflated. I clarified in an edit, but basically, this means that not all of those 6% of requests get the full benefits of defer, because of that async/defer anti-pattern.

Considering that `defer` provides us with the best loading performance (by ensuring downloading the script happens in parallel to other work, and execution waits until after the page can be displayed), we would hope to see that percentage a bit higher. In fact, as it is that 6.0% is a bit misleading.

Back when supporting IE8 and IE9 was more common, it was relatively common to use _both_ the `async` and `defer` attributes. With both attributes in place, any browser supporting both will use `async`. IE8 and IE9, which don't support `async` will fall back to defer.

> {# TODO(authors): In the Twitter thread that discussed this, it actually spurred the Jetpack folks to push a fix! Might be worth mentioning :) #}

Author: @tkadlec totally optional

Reviewer 3: Great little story! Added. :)

Author: Thanks! Could you open a new PR with your changes?

  sql_file="compression_method.sql"
) }}

> 85% of all JavaScript requests have some level of network compression applied. Gzip makes up the majority of that, with 65% of scripts having gzip compression applied compared to 20% for Brotli (br). While the percentage of Brotli (which is more effective than gzip) is low compared to its browser support, it's trending in the right direction, increasing by 5 percentage points in the last year.

Reviewer 4: Hmm, I thought that spelling "Gzip" with a capital letter was correct. Because during the Compression chapter edits this spelling was used (see https://github.com/HTTPArchive/almanac.httparchive.org/pull/1688/commits/3b475a13c9fab10679a6b11158e6b8dfe8c9711c). Just want to achieve consistency between chapters.

Author: Good catch. I did a search and found most usage to be lowercase. We should be consistent between chapters but now I'm not sure which is more correct.

Reviewer 2: GNU says Gzip: https://www.gnu.org/software/gzip/

I say Gzip for format and `gzip` for command and file eztension.

It's important that we think critically about the tools we use to build the web and what the trade-offs are, so it makes sense to look closely at what we see in use today.

### Libraries

HTTP Archive uses [Wappalyzer](./methodology#wappalyzer) to detect technologies in use on a given page. Wappalazyer tracks both JavaScript libraries (think of these as a collection of snippets or helper functions to ease development, like jQuery) and JavaScript frameworks (these are more likely scaffolding and provide templating and structure, like React).

The popular libraries in use are largely unchanged from last year, with jQuery continuing to dominate usage and only one of the top 21 libraries falling out (lazy.js, replaced by DataTables). In fact, even the percentages of the top libraries has barely changed from last year.

> {# TODO(analysts): table? showing rank, library, percentage and last years rank #}

Author: This requires pulling 2019 data so a bit more involved. I wonder if the figure below gets the job done. @tkadlec LMK.

Reviewer 3: Sorry...what figure is that?

Author: https://almanac.httparchive.org/en/2020/javascript?#fig-21

(The changes in this PR were just deployed to production)

Reviewer 3: Gotcha. I think that's fine. A comparison is kinda boring anyway with the lack of change.

Both are very sound guesses, and it seems the situation hasn't changed much on either front.

In fact, the dominance of jQuery is supported even further when you stop to consider that, of the top 10 libraries, 6 of them are either jQuery or require jQuery in order to be used: jQuery UI, jQuery Migrate, FancyBox, Lightbox and Slick.

### Frameworks

When we look at the frameworks, we also don't see much of a dramatic change in terms of adoption in the main frameworks that were highlighted last year. Vue.js has seen a significant increase, and AMP grew a bit, but most of them are more or less where they were a year ago.

{# TODO(analysts): Compare same frameworks from last year's chapter to this year in bar chart? #}

> {# TODO(authors): Elaborate on what the detection issue was. #}

Author: @tkadlec

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vRn1IaMxnTl0jhdC-C-vC5VLN_boJfLAaOfGJ968IalK1vPc8-dz0OkVmNY0LjMxZ6BIwSRB7xtRmIE/pubchart?oid=1758266664&format=interactive",
  sheets_gid="160832134",
  sql_file="main_thread_time_frameworks.sql"
) }}

As you would expect, there's a gap for all tools in use due to the lower processing power of the [emulated Moto G4](methodology#webpagetest). Ember and Polymer seem to jump out as particularly egregious examples, while tools like RxJS and Mustache vary only minorly from desktop to mobile.

## What's the impact?

We have a pretty good picture now of how much JavaScript we use, where it comes from and what we use it for. While that's interesting enough on its own, the real kicker is the "so what?" What impact does all this script actually have on the experience of our pages?

> The first thing we should consider is what happens with all that JavaScript once its been downloaded. Downloading is only the first part of the JavaScript journey. The browser still has to parse all that script, compile it, and eventually execute it. While browsers are increasingly finding ways to offload

Reviewer: > While browsers are increasingly finding ways to offload

This sentence continues to be incomplete, I remember reporting it earlier.

Author: Yes, just got to that part of the edit and left a TODO

Reviewer 3: Fixed! Sorry for missing that the first time around @ibnesayeed!
